File: stripe/src/test/java/com/stripe/android/ErrorParserTest.java
package com.stripe.android;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

/**
 * Test class for {@link ErrorParser}.
 */
public class ErrorParserTest {

    private static final String RAW_INVALID_REQUEST_ERROR = "" +
            "{\n" +
            "  \"error\": {\n" +
            "    \"message\" : \"The Stripe API is only accessible over HTTPS.  " +
            "Please see <https://stripe.com/docs> for more information.\",\n" +
            "    \"type\": \"invalid_request_error\"\n" +
            "  }\n" +
            "}";

    private static final String RAW_INCORRECT_FORMAT_ERROR = "" +
            "{\n" +
            "    \"message\" : \"The Stripe API is only accessible over HTTPS.  " +
            "Please see <https://stripe.com/docs> for more information.\",\n" +
            "    \"type\": \"invalid_request_error\"\n" +
            "}";

    private static final String RAW_ERROR_WITH_ALL_FIELDS = "" +
            "{\n" +
            "  \"error\": {\n" +
            "    \"code\": \"code_value\",\n" +
            "    \"param\": \"param_value\",\n" +
            "    \"charge\": \"charge_value\",\n" +
            "    \"decline_code\": \"card_declined\",\n" +
            "    \"message\": \"Your card was declined.\",\n" +
            "    \"type\": \"invalid_request_error\"\n" +
            "  }\n" +
            "}";

    @Test
    public void parseError_withInvalidRequestError_createsCorrectObject() {
        final StripeError parsedStripeError =
                ErrorParser.parseError(RAW_INVALID_REQUEST_ERROR);
        String errorMessage = "The Stripe API is only accessible over HTTPS.  " +
                "Please see <https://stripe.com/docs> for more information.";
        assertEquals(errorMessage, parsedStripeError.getMessage());
        assertEquals("invalid_request_error", parsedStripeError.getType());
        assertEquals("", parsedStripeError.getParam());
    }

    @Test
    public void parseError_withNoErrorMessage_addsInvalidResponseMessage() {
        final StripeError badStripeError =
                ErrorParser.parseError(RAW_INCORRECT_FORMAT_ERROR);
        assertEquals(ErrorParser.MALFORMED_RESPONSE_MESSAGE, badStripeError.getMessage());
        assertNull(badStripeError.getType());
    }

    @Test
    public void parseError_withAllFields_parsesAllFields() {
        final StripeError error = ErrorParser.parseError(RAW_ERROR_WITH_ALL_FIELDS);
        assertEquals("code_value", error.getCode());
        assertEquals("param_value", error.getParam());
        assertEquals("charge_value", error.getCharge());
        assertEquals("Your card was declined.", error.getMessage());
        assertEquals("card_declined", error.getDeclineCode());
        assertEquals("invalid_request_error", error.getType());
    }
}

File: stripe/src/test/java/com/stripe/android/ErrorParserTest.kt
package com.stripe.android

import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * Test class for [ErrorParser].
 */
class ErrorParserTest {

    @Test
    fun parseError_withInvalidRequestError_createsCorrectObject() {
        val (type, message, _, param) = ErrorParser.parseError(RAW_INVALID_REQUEST_ERROR)
        val errorMessage = "The Stripe API is only accessible over HTTPS.  " + "Please see <https://stripe.com/docs> for more information."
        assertEquals(errorMessage, message)
        assertEquals("invalid_request_error", type)
        assertEquals("", param)
    }

    @Test
    fun parseError_withNoErrorMessage_addsInvalidResponseMessage() {
        assertEquals(ErrorParser.MALFORMED, ErrorParser.parseError(RAW_INCORRECT_FORMAT_ERROR))
    }

    @Test
    fun parseError_withAllFields_parsesAllFields() {
        val (type, message, code, param, declineCode, charge) = ErrorParser.parseError(RAW_ERROR_WITH_ALL_FIELDS)
        assertEquals("code_value", code)
        assertEquals("param_value", param)
        assertEquals("charge_value", charge)
        assertEquals("Your card was declined.", message)
        assertEquals("card_declined", declineCode)
        assertEquals("invalid_request_error", type)
    }

    companion object {
        private val RAW_INVALID_REQUEST_ERROR =
            """
            {
                "error": {
                    "message": "The Stripe API is only accessible over HTTPS.  Please see <https://stripe.com/docs> for more information.",
                    "type": "invalid_request_error"
                }
            }
            """.trimIndent()

        private val RAW_INCORRECT_FORMAT_ERROR =
            """
            {
                "message": "The Stripe API is only accessible over HTTPS.  Please see <https://stripe.com/docs> for more information.",
                "type": "invalid_request_error"
            }
            """.trimIndent()

        private val RAW_ERROR_WITH_ALL_FIELDS =
            """
            {
                "error": {
                    "code": "code_value",
                    "param": "param_value",
                    "charge": "charge_value",
                    "decline_code": "card_declined",
                    "message": "Your card was declined.",
                    "type": "invalid_request_error"
                }
            }
            """.trimIndent()
    }
}
